lib/Conversion/TorchOnnxToTorch/OnnxLstmExpander.cpp

```diff
@@ -0,0 +1,479 @@
+
+  Value cstOne = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 1));
+
+  // Apply linear/matmul for each gate separately
+  // names are consistent with ONNX LSTM documentation
+  Value i_x = rewriter.create<Torch::AtenLinearOp>(loc, HType, Xt, W_i, Wb_i);
+  Value i_h =
+      rewriter.create<Torch::AtenLinearOp>(loc, HType, H_prev, R_i, Rb_i);
+  Value i =
+      rewriter.create<Torch::AtenAddTensorOp>(loc, HType, i_x, i_h, cstOne);
+  Value i_act = createActivationByName(rewriter, loc, activations[0], i);
+
+  Value o_x = rewriter.create<Torch::AtenLinearOp>(loc, HType, Xt, W_o, Wb_o);
+  Value o_h =
+      rewriter.create<Torch::AtenLinearOp>(loc, HType, H_prev, R_o, Rb_o);
+  Value o =
+      rewriter.create<Torch::AtenAddTensorOp>(loc, HType, o_x, o_h, cstOne);
+  Value o_act = createActivationByName(rewriter, loc, activations[0], o);
+
+  Value f_x = rewriter.create<Torch::AtenLinearOp>(loc, HType, Xt, W_f, Wb_f);
+  Value f_h =
+      rewriter.create<Torch::AtenLinearOp>(loc, HType, H_prev, R_f, Rb_f);
+  Value f =
+      rewriter.create<Torch::AtenAddTensorOp>(loc, HType, f_x, f_h, cstOne);
+  Value f_act = createActivationByName(rewriter, loc, activations[0], f);
+
+  Value ct_x = rewriter.create<Torch::AtenLinearOp>(loc, HType, Xt, W_c, Wb_c);
+  Value ct_h =
+      rewriter.create<Torch::AtenLinearOp>(loc, HType, H_prev, R_c, Rb_c);
+  Value ct =
+      rewriter.create<Torch::AtenAddTensorOp>(loc, HType, ct_x, ct_h, cstOne);
+  Value ct_act = createActivationByName(rewriter, loc, activations[1], ct);
+
+  Value C_forget =
+      rewriter.create<Torch::AtenMulTensorOp>(loc, HType, f_act, C_prev);
+  Value C_input =
+      rewriter.create<Torch::AtenMulTensorOp>(loc, HType, i_act, ct_act);
+  Value C_new = rewriter.create<Torch::AtenAddTensorOp>(loc, HType, C_forget,
+                                                        C_input, cstOne);
+
+  Value C_new_act =
+      createActivationByName(rewriter, loc, activations[2], C_new);
+  Value H_new =
+      rewriter.create<Torch::AtenMulTensorOp>(loc, HType, o_act, C_new_act);
+
+  return std::make_pair(H_new, C_new);
+}
+
+std::tuple<Value, Value, Value> lstm_layer( // returns Y, Y_h, Y_c
+    OpBinder binder, ConversionPatternRewriter &rewriter, Value X,
+    // X shape [seq_length, batch_size, input_size]
+    Value initial_h,
+    // =hidden_state shape [batch_size, hidden_size]
+    Value initial_c,
+    // initial_c shape [batch_size, hidden_size]
+    Value W,
+    // W shape [hidden_size*4, input_size]
+    Value Wb,
+    // Wb shape [hidden_size*4]
+    Value R,
+    // R shape [hidden_size*4, hidden_size]
+    Value Rb,
+    // Rb shape [hidden_size*4]
+    SmallVector<std::string> activations) {
+
+  Location loc = binder.getLoc();
+  // these names are snake_case for consistency with onnx.LSTM documentation
+  int64_t seq_len = X.getType().cast<Torch::ValueTensorType>().getSizes()[0];
```

Comment:
Rather than getting / casting the type each line it is better to just do it once and reuse.

```
auto xTy = x.getType().cast<Torch::ValueTensorType>();
```